File: tests/lua/opt_constant_fold.ml
external val ignore : 'a -> () = "function(x) end"

external val ( +  ) : int -> int -> int = "%int.add"
external val ( -  ) : int -> int -> int = "%int.sub"
external val ( *  ) : int -> int -> int = "%int.mul"
external val ( ** ) : int -> int -> int = "%int.pow"
external val ( /  ) : int -> int -> int = "%int.div"
external val ( == ) : int -> int -> bool = "%int.eq"
external val ( <= ) : int -> int -> bool = "%int.le"
external val ( <  ) : int -> int -> bool = "%int.lt"

external val ( +.  ) : float -> float -> float = "%float.add"
external val ( -.  ) : float -> float -> float = "%float.sub"
external val ( *.  ) : float -> float -> float = "%float.mul"
external val ( **. ) : float -> float -> float = "%float.pow"
external val ( /.  ) : float -> float -> float = "%float.div"
external val ( ==. ) : float -> float -> bool = "%float.eq"
external val ( <=. ) : float -> float -> bool = "%float.le"
external val ( <.  ) : float -> float -> bool = "%float.lt"

external val ( ^ ) : string -> string -> bool = "%string.concat"
external val ( ==? ) : string -> string -> bool = "%string.eq"
external val ( <=? ) : string -> string -> bool = "%string.le"
external val ( <?  ) : string -> string -> bool = "%string.lt"


let () = ignore (fun { i, s }->
  { int_add_c = 2 + 3
  , int_sub_c = 2 - 3
  , int_mul_c = 2 * 3
  , int_div_c = 6 / 2
  , int_exp_c = 2 ** 3
  , int_eq_c  = 2 == 3
  , int_lt_c  = 2 < 3
  , int_le_c  = 2 <= 3

  , int_add_l = i + 0
  , int_add_r = 0 + i
  , int_sub_l = i - 0
  , int_mul1_l = i * 1
  , int_mul1_r = 1 * i
  , int_mul_l = i * 0
  , int_mul_r = 0 * i
  , int_div_l = i / 1

  , float_add_c = 2.0 +. 3.0
  , float_sub_c = 2.0 -. 3.0
  , float_mul_c = 2.0 *. 3.0
  , float_div_c = 6.0 /. 2.0
  , float_exp_c = 2.0 **. 3.0
  , float_eq_c  = 2.0 ==. 3.0
  , float_lt_c  = 2.0 <. 3.0
  , float_le_c  = 2.0 <=. 3.0

  , str_con_c = "foo" ^ "bar"
  , str_eq_c = "foo" ==? "bar"
  , str_le_c = "foo" <=? "bar"
  , str_lt_c = "foo" <? "bar"

  , str_con_l = s ^ ""
  , str_con_r = "" ^ s

  , int_eq_u = i == i
  , str_eq_u = s ==? s
  })

File: tests/lua/opt_constant_fold.lua
do
  local ignore = function(x)  end
  ignore(function(tmp)
    local i, s = tmp.i, tmp.s
    return {
      int_add_c = 5,
      int_sub_c = -1,
      int_mul_c = 6,
      int_div_c = 3.0,
      int_exp_c = 8,
      int_eq_c = false,
      int_lt_c = true,
      int_le_c = true,
      int_add_l = i,
      int_add_r = i,
      int_sub_l = i,
      int_mul1_l = i,
      int_mul1_r = i,
      int_mul_l = 0,
      int_mul_r = 0,
      int_div_l = i,
      float_add_c = 5.0,
      float_sub_c = -1.0,
      float_mul_c = 6.0,
      float_div_c = 3.0,
      float_exp_c = 8.0,
      float_eq_c = false,
      float_lt_c = true,
      float_le_c = true,
      str_con_c = "foobar",
      str_eq_c = false,
      str_le_c = false,
      str_lt_c = false,
      str_con_l = s,
      str_con_r = s,
      int_eq_u = true,
      str_eq_u = true
    }
  end)
end
